pyro/generic/generic.py

```diff
@@ -27,6 +31,39 @@ def __getattribute__(self, name):
         return getattr(module, name)
 
 
+def seed(fn=None, rng_key=None):
```

Comment:
This should be provided by each backend, otherwise there is no way for backends to override this. I believe it should live either in poutine or primitives (and be a top-level `pyro.seed` decorator).

Reply:
This currently only affects the NumPyro backend and should be a NoOp for other backends (that's the reason for the if block below), which are free to instead call their own mechanism to set the random seed. This was motivated by your previous suggestion to redesign keeping the flow for Pyro the same (maybe I misunderstood your earlier comment?). I also think it will be better for this to be in `pyro.poutine`. 

Do you have any suggestions as to the behavior of that poutine? Should it just call `pyro.set_rng_seed` in `__init__` and have no other effect? Note that this needs to be a poutine in NumPyro since it has to intercept every `sample` statement. With that design, we can just remove this function and also `generic` from dispatch and let `handlers.seed` determine how each backend likes to set the rng seed.

Reply:
I believe you and @eb8680 and I agreed on a new handler that has no effect in Pyro:
```py
#in pyro/poutine/handlers.py
def model(fn=None, **kwargs):
    """
    Decorator to declare a Pyro model. This is optional and has no effect in the
    standard Pyro backend, but this may be required for alternative backends, e.g.
    in the NumPyro backend this may install default handlers.
    """
    if fn is None:
        return model
    return fn
```
I would also be happy for this to live in primitives.py, so we can wrap with `@pyro.model`. @eb8680 do you have an opinion?

Reply:
Ahh..I see what you meant now, I had this model decorator in generic earlier, but it will be better if this resides in `primitives`. I would prefer if this was simply a `seed` context manager in `pyro.primitives`, so that there is a way to set the rng seed in a backend agnostic way. 

```
with seed(rng_key=1):
  ...
```

For Pyro this will just call `pyro.set_rng_seed(..)`. Otherwise, we will need to separately call `pyro.set_rng_seed` for the Pyro backend and use the pattern below for NumPyro, which requires detecting what backend we are on. From the examples and tests, I think the contextmanager use is more important rather than the decorator pattern, since the latter will give the same samples from the model, but both can be supported if this is a poutine.

Reply:
A seed context manager in primitives sounds great, thanks for explaining.

Reply:
This is added to `handlers` now. Could you take a look and see if that looks okay? One advantage to that is that we can use it both as a context manager and decorator and the effect is the same across all backends.